source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc

```diff
@@ -53,16 +53,16 @@ Http::FilterHeadersStatus ProxyFilter::decodeHeaders(Http::RequestHeaderMap& hea
   }
   cluster_info_ = cluster->info();
 
-  auto& resource = cluster_info_->resourceManager(route_entry->priority()).pendingRequests();
-  if (!resource.canCreate()) {
-    ENVOY_STREAM_LOG(debug, "pending request overflow", *decoder_callbacks_);
-    cluster_info_->stats().upstream_rq_pending_overflow_.inc();
-    decoder_callbacks_->sendLocalReply(
+  auto& pending_requests = config_->cache().onDnsRequest(route_entry, cluster_info_);
+  if (pending_requests.canCreate()) {
```

Comment:
Per my other comment if you return an optional reference you don't need to call can create here again which is itself a race condition.